File: Day007_2025/Q007_Leetcode.py
from typing import List
class TrieNode:
    def __init__(self):
        self.frequency = 0
        self.childNodes = {}

class Solution:
    def stringMatching(self, words: List[str]) -> List[str]:
        # Q. String Matching in an Array
        root = TrieNode()
        matching_words = []

        # Insert substrings into the Trie
        for word in words:
            for start_index in range(len(word)):
                self.insert_word(root, word[start_index:])

        # Check if each word is a substring of another word
        for word in words:
            if self.is_substring(root, word):
                matching_words.append(word)

        return matching_words

    def insert_word(self, root: TrieNode, word: str):
        current_node = root
        for char in word:
            if char not in current_node.childNodes:
                current_node.childNodes[char] = TrieNode()
            current_node = current_node.childNodes[char]
            current_node.frequency += 1

    def is_substring(self, root: TrieNode, word: str) -> bool:
        current_node = root
        for char in word:
            if char not in current_node.childNodes:
                return False
            current_node = current_node.childNodes[char]
        return current_node.frequency > 1


# Example Usage
if __name__ == "__main__":
    solution = Solution()
    words = ["mass", "as", "hero", "superhero"]
    print(solution.stringMatching(words))  # Output: ['as', 'hero']

File: Day007_2025/Q007_Leetcode.java
import java.util.*;

class Q007_Leetcode {
    class TrieNode {
        int frequency;
        Map<Character, TrieNode> childNodes;

        TrieNode() {
            this.frequency = 0;
            this.childNodes = new HashMap<>();
        }
    }

    public List<String> stringMatching(String[] words) {
        // Q. String Matching in an Array
        TrieNode root = new TrieNode();
        List<String> matchingWords = new ArrayList<>();

        // Insert substrings into the Trie
        for (String word : words) {
            for (int startIndex = 0; startIndex < word.length(); startIndex++) {
                insertWord(root, word.substring(startIndex));
            }
        }

        // Check if each word is a substring of another word
        for (String word : words) {
            if (isSubstring(root, word)) {
                matchingWords.add(word);
            }
        }

        return matchingWords;
    }

    private void insertWord(TrieNode root, String word) {
        TrieNode currentNode = root;
        for (char c : word.toCharArray()) {
            currentNode.childNodes.putIfAbsent(c, new TrieNode());
            currentNode = currentNode.childNodes.get(c);
            currentNode.frequency++;
        }
    }

    private boolean isSubstring(TrieNode root, String word) {
        TrieNode currentNode = root;
        for (char c : word.toCharArray()) {
            if (!currentNode.childNodes.containsKey(c)) {
                return false;
            }
            currentNode = currentNode.childNodes.get(c);
        }
        return currentNode.frequency > 1;
    }

    public static void main(String[] args) {
        Q007_Leetcode solution = new Q007_Leetcode();
        String[] words = {"mass", "as", "hero", "superhero"};
        System.out.println(solution.stringMatching(words)); // Output: [as, hero]
    }
}
